pkg/kubelet/util/boottime_util_linux.go

```diff
@@ -21,13 +21,39 @@ package util
 
 import (
 	"fmt"
+	"os"
+	"strconv"
+	"strings"
 	"time"
 
 	"golang.org/x/sys/unix"
+	"k8s.io/klog/v2"
 )
 
-// GetBootTime returns the time at which the machine was started, truncated to the nearest second
+// GetBootTime returns the time at which the machine was started, truncated to the nearest second.
+// It uses /proc/uptime first, which is more accurate, and falls back to the less accurate
+// unix.Sysinfo if /proc/uptime failed.
 func GetBootTime() (time.Time, error) {
+	currentTime := time.Now()
+	raw, err := os.ReadFile("/proc/uptime")
+	if err != nil {
+		klog.Errorf("error getting system uptime: %s", err)
+		return getBootTimeWithSysinfo()
+	}
+	uptimeStr, _, found := strings.Cut(string(raw), " ")
+	if !found {
+		klog.Errorf("unexpected value from /proc/uptime: %s", raw)
+		return getBootTimeWithSysinfo()
+	}
+	uptime, err := strconv.ParseFloat(uptimeStr, 64)
+	if err != nil {
+		klog.Errorf("error parsing uptime %s: %s", uptimeStr, err)
+		return getBootTimeWithSysinfo()
```

Comment:
this sounds very brittle can we have unit test for this?

Reply:
hmm yeah, the existing unit test only cares about getting a result so with this fall back behavior we don't know if the result is produced with /proc/uptime or unix.Sysinfo. How about further separating them into `func getBootTimeWithSysinfo()` and `func getBootTimeWithProcUptime()` so we can test them separately. Then in the entry point `func GetBooTime()` we just call out to the 2 methods. WDYT?

Reply:
k I updated the commit as described, so marking as resolved. Let me know if you have further comments-)